apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRDASource.java

```diff
@@ -0,0 +1,152 @@
+package gov.cms.bfd.pipeline.rda.grpc.source;
+
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.annotations.VisibleForTesting;
+import gov.cms.bfd.pipeline.rda.grpc.PreAdjudicatedClaim;
+import gov.cms.bfd.pipeline.rda.grpc.ProcessingException;
+import gov.cms.bfd.pipeline.rda.grpc.RDASink;
+import gov.cms.bfd.pipeline.rda.grpc.RDASource;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * General RDASource implementation that delegates actual service call and result mapping to another
+ * class.
+ *
+ * @param <T> type of objects returned by the gRPC service
+ */
+public class GrpcRDASource<T> implements RDASource<PreAdjudicatedClaim> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(GrpcRDASource.class);
+  public static final String CALLS_METER =
+      MetricRegistry.name(GrpcRDASource.class.getSimpleName(), "calls");
+  public static final String RECORDS_RECEIVED_METER =
+      MetricRegistry.name(GrpcRDASource.class.getSimpleName(), "recordsReceived");
+  public static final String RECORDS_STORED_METER =
+      MetricRegistry.name(GrpcRDASource.class.getSimpleName(), "recordsStored");
+  public static final String BATCHES_METER =
+      MetricRegistry.name(GrpcRDASource.class.getSimpleName(), "batches");
+
+  private final GrpcStreamCaller.Factory<T> callerFactory;
+  private final Clock clock;
+  private final Meter callsMeter;
+  private final Meter recordsReceivedMeter;
+  private final Meter recordsStoredMeter;
+  private final Meter batchesMeter;
+  private ManagedChannel channel;
+
+  public GrpcRDASource(
+      Config config, GrpcStreamCaller.Factory<T> callerFactory, MetricRegistry appMetrics) {
+    this(
+        ManagedChannelBuilder.forAddress(config.host, config.port)
+            .usePlaintext()
+            .idleTimeout(config.maxIdle.toMillis(), TimeUnit.MILLISECONDS)
+            .build(),
+        callerFactory,
+        Clock.systemDefaultZone(),
+        appMetrics);
+  }
+
+  @VisibleForTesting
+  GrpcRDASource(
+      ManagedChannel channel,
+      GrpcStreamCaller.Factory<T> callerFactory,
+      Clock clock,
+      MetricRegistry appMetrics) {
+    this.callerFactory = callerFactory;
+    this.channel = channel;
+    this.clock = clock;
+    callsMeter = appMetrics.meter(CALLS_METER);
+    recordsReceivedMeter = appMetrics.meter(RECORDS_RECEIVED_METER);
+    recordsStoredMeter = appMetrics.meter(RECORDS_STORED_METER);
+    batchesMeter = appMetrics.meter(BATCHES_METER);
+  }
+
+  @Override
+  public int retrieveAndProcessObjects(
+      int maxToProcess, int maxPerBatch, Duration maxRunTime, RDASink<PreAdjudicatedClaim> sink)
```

Comment:
NULL maxRuntime would cause an exception.  0 maxToProcess would exit the loop immediately.  I'd rather not build in overrides for misconfiguration at the lowest levels.  We could validate the values here but the caller should already have done so and I'd prefer not to do the same validation in multiple places.

Reply:
Let me provide more context, if we did want to run the process indefinitely, what values would we pass and/or how would we configure the job?

Reply:
This would do in practice.

````java
maxToProcess = Integer.MAX_VALUE;
maxRuntime = Duration.ofMilliseconds(Long.MAX_VALUE);
````

Reply:
> This would do in practice.

Like I noted in the other PR, I think I'd recommend removing that functionality altogether. Operationally, it has bad characteristics if we actually use it, and if we're not using it, it's just extra code we don't need.

Reply:
When this was developed I saw a single thread being used in PipelineManager.  In that context maxRunTime was intended to allow the job to time slice over a single thread without hogging it.  In the past I have found a setting like maxToProcess can be useful when testing or to validate the service is working without risking full blown processing (quick pre-flight check).  They could both be useful in the context of an integration test as well.  It seems I'm the only one who values this small feature.

Once the new orchestration code (in time for this sprint?) is in place this code will need to be refactored anyway.  I could make any adjustments in the next PR which will call the actual RDA 0.1 MVP service.